README.md
# InBrain Surveys
Survey library to monetize your mobile app, provided by inBrain.ai

## Requirements
This SDK is targeted to the following tools:
- XCode 12
- iOS 11.0;
- CocoaPods 1.10+
- Swift 5
- React Native >=0.63.0

## Installation

Install and link the module:

`$ npm install inbrain-surveys --save`

### Extra steps iOS

Do not forget to use Cocoapods 1.9 in your project. Handling of xcframeworks isn't well supported in the previous versions. 

Set the framework 'Target Membership' to `inbrain-surveys` as below:

![Framework Target Membership](https://i.ibb.co/N2ntq0P/target-membership.png)

### Extra steps Android
Add jitpack repository you your gradle configuration `android/build.gradle > allprojects > repositories`

```
maven { 
    url 'https://jitpack.io' 
}
```

## Usage
```javascript
import inbrain from 'inbrain-surveys';
```

For a fully functional example, please refer to this [demo app](https://github.com/inbrainai/react-native) using the SDK.

Available functions:
### Initialise the SDK
```javascript
inbrain.init(apiClientId: string, apiSecret: string, options?: InitOptions) : Promise<void>
```
* `apiClientId`: The client ID obtained from your account manager
* `apiSecret`: The client secret obtained from your account manager.
* `options`: [Optional] Options. Possible options:
    * `language`: Accepted languages: `de-de`, `en-au`, `en-ca`, `en-gb`, `en-in`, `en-us`, `es-es`, `es-mx`, `es-us`, `fr-ca`, `fr-fr`, `fr-br` (case sensitive). Default to the device's locale language.
    * `sessionUid`: Value to track each session of inBrain use from a specific userID
    * `isS2S`: If the SDK runs in Server To Server mode. Default `false`
    * `userId`: The unique string value that differentiates each user within their app when initializing inBrain (Example: an email, a username). Default `''`
    * `title`: The surveys view title. Default 'inBrain.ai Surveys'
    * `navigationBar`: The navigation bar configuration
        * `backgroundColor`: The navigation bar background color (hexadecimal string color, e.g #FF0000)
        * `buttonsColor`: The navigation bar buttons color (hexadecimal string color, e.g #E7F722)
        * `titleColor`: The navigation bar title color (hexadecimal string color, e.g #FF0404)
        * `hasShadow`: The navigation bar bottom border (boolean)
    * `statusBar`: The status bar configuration (boolean)
        * `lightStatusBar`: The status bar text and icons color (true for white, false for black). On iOS, you'll need to set `View controller-based status bar appearance` bar appearance to YES for this option to work.
    * `dataPoints`: A dictionary of keys and values to provide inBrain profiler data for custom profiler user experience (Example: `{ age : “23”, gender : “female” }`)

Note: This method need to be called prior calling all the other methods. 

### Change the session parameters
```javascript
inbrain.setSessionParameters(sessionUid: string, dataPoints: {}) : Promise<void>
```
* sessionUid the session identifiers
* dataPoints datapoints

### Show the surveys webview
```javascript
inbrain.showSurveys() : Promise<void>
```

### Get the rewards (Useful for server less app)
```javascript
inbrain.getRewards() : Promise<InBrainReward[]>
```

### Confirm a list of rewards (Useful for server less app)
```javascript
inbrain.confirmRewards(rewards: InBrainReward[]) : Promise<void>
```
* rewards: List of rewards to confirm

### Check if Native Surveys are available
```javascript
inbrain.checkSurveysAvailable() : Promise<boolean>
```

### Get the Native Surveys available
```javascript
inbrain.getNativeSurveys(placementId?: string) : Promise<InBrainNativeSurveys[]>
```
* placementId: an optional placement identifier

### Show a Native Survey
```javascript
inbrain.showNativeSurvey(id: string, placementId?: string) : Promise<void>
```
* placementId: an optional placement identifier

### On webview dismissed
```javascript
inbrain.setOnCloseListener(callback: () => void) 
```
* callback: callback to perform when it happens

Note: Calling this method multiple times will override the previous listener.

### On webview dismissed from page
```javascript
inbrain.setOnCloseListenerFromPage(callback: () => void) 
```
* callback: callback to perform when it happens

Note: Calling this method multiple times will override the previous listener.

## Troubleshoots
### [BUILD TIME] Several errors similar to `Bad receiver 'int *'`, `Unkown type name 'InBrain'`...
First check that the framework target membership is correctly set. If the problem still occurs, and you are trying to create a release build with a simulator as target (and you are using XCode 12), please add `arm64` in the Excluded Architecture of your project and 'inbrain-surveys' pods (see [this post](https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios) for detailed information)

### [BUILD TIME] 'InBrainSurveys_SDK_Legacy/InBrainSurveys_SDK_Legacy-Swift.h' file not found
This problem usually happens if the framework is not set in the Embedded Binaries, or if the framework doesn't have the target set to `inbrain-surveys` See above for set up.
Clean and build the project after changes.

### [RUNTIME] Library not loaded: @rpath/libswiftCore.dylib
This problem can happen if your project doesn't have the Swift standard libraries included. Set the 'Always Embed Swift Standard Libraries' to yes in your target to fix it.
Clean and build the project after changes.
This problem also consistently appears when usinx XCode10

### [RUNTIME - Release scheme] dependent dylib '@rpath/InBrainSurveys_SDK_Swift.framework/InBrainSurveys_SDK_Swift' not found 
This problem happen with previous version of Cocoapods and XCode. Try updating to Cocoapods 1.10.x, and if it still doesn't work, also upgrade to XCode12

### [DEVELOPMENT] When running 'pod install' in APP: "Specs satisfying the `inbrain-surveys (from `../node_modules/inbrain-surveys`)` dependency were found, but they required a higher minimum deployment target."
 => Modify the platform version in 'APP/ios/Podfile'

 ### [DEVELOPMENT] When building in XCode: No matching function for call to 'RCTBridgeModuleNameForClass' 
Update from RN0.62 to RN0.63

 ### [DEVELOPMENT] When building in XCode: problem with not finding switch libraries
https://github.com/facebook/react-native/issues/31733#issuecomment-931824830


CHANGELOG.md
## [2.3.2] 2022-10-14
### Changed
- Updated RN version from `0.63` to `0.70.2`
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.8.7`
- Fix library compatibility issue with the latest version of react native

## [2.3.0] 2022-03-07
### Changed
- Updated RN version from `0.62` to `0.63`
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.8.4`

## [2.2.4] 2020-11-29
### Changed
- Added two new properties in InBrainNativeSurveys: currencySale and multiplier
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.21` (from `1.0.120`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.8.2`

## [2.2.3] 2020-11-29
### Fixed
- Fixed 2.2.2 crash related to NativeSurveysDelegate on iOS

## [2.2.2] 2020-11-27
### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.20` (from `1.0.19`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.8.1`

## [2.2.1] SKIPPED - was unpublished from npmjs (published by mistake)

## [2.2.0] 2020-10-09
### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.19` (from `1.0.16`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.8.0`
- Added `placementId` optional parameters to methods `getNativeSurveys` and `showNativeSurvey`
- Added method `setSessionParameters` to update parameters without having to call `init`

## [2.1.4] 2020-01-27
### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.16` (from `1.0.15`)

## [2.1.3] 2020-01-23
### Changed
- Merged v1.1.4 and v1.1.5 into 2.1.x
- `showSurveys` and `showNativeSurveys` are now executed from the UI Thread in Android

## [2.1.2] 2020-01-20
### Changed
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.5.2`

## [2.1.1] 2020-01-16
### Bug Fix
- Fixed and improved error messages on iOS
### Changed
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.5.1`

## [2.1.0] 2020-01-14
### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.15` (from `1.0.14`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.5.0`

## [2.0.2] 2020-12-31
### Changed
- Added some Javascript side validation
### Merged
- Merged 1.1.4 into 2.0.0: Improve handling of null Activity.

## [2.0.1] 2020-12-21
### Merged
- Merged 1.1.3 into 2.0.0: Internals of Android bridge. Prevent app from crashing if problem during init call.

## [2.0.0] 2020-12-12
### Changed
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.4.12`
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.14` (from `1.0.9`)

### Changed
- Moved options `titleColor` and `navbarColor`
- Added two different kind of options `navigationBar` and `statusBar` to have more fine-grained control over navigation and status bars styling
## [1.1.5] 2021-01-22
### Changed
- `showSurveys` and `showNativeSurveys` are now executed from the UI Thread in Android

## [1.1.5] 2021-01-23
### Changed
- `showSurveys` and `showNativeSurveys` are now executed from the UI Thread in Android

## [1.1.4] 2020-12-31
### Changed
- Improve handling of null Activity (bridge method is being re-called at most 10 times before throwing, with 50ms intervals) in Android

## [1.1.3] 2020-12-18
### Changed
- Internals of Android bridge. Prevent app from crashing if problem during init call.

## [1.1.2] 2020-11-20
### Added
- Added option 'titleColor' which allows to modify the webview title color

## [1.1.1] 2020-11-19
### Bug Fix
- Fix onClose listener not triggered when closing the webview displaying a native survey
### Changed
- Simplified ObjC code and removed unnecessary view controller 

## [1.1.0] 2020-11-13
### Added
- Added CHANGELOG.md
- Added methods `checkSurveysAvailable`, `getNativeSurveys`, and `showNativeSurvey` methods

### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.9` (from `1.0.3`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to version `1.4.6`

## [1.0.8] - 2020-09-02
### Changed
- Updated iOS framework `InBrainSurveys_SDK_Swift.xcframework`

## [1.0.7] - 2020-07-28
### Added
- Added attribute `isS2S` in `InitOptions` with default value `false`

### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `1.0.3` (from `0.1.25`)
- Updated iOS framework `InBrainSurveys_SDK_Legacy.xcframework` to `InBrainSurveys_SDK_Swift.xcframework`
- Changed bridge internal method names and prototypes
- Set default value for `userId` attribute in `InitOptions` to `''` 

### Removed
- Removed attribute `production` in `InitOptions`

## [1.0.6] - 2020-06-17
### Changed
- Updated Android framework `com.github.inbrainai:sdk-android` to version `0.1.25` (from `0.1.9`)

## [1.0.5] - 2020-06-01

## [1.0.4] - 2020-05-31

## [1.0.3] - 2020-05-21

## [1.0.2] - 2020-05-14

## [1.0.1] - 2020-04-21

## [1.0.0] - 2020-02-13
### Added
- Initial SDK version


ios/InBrainSurveys.m
#import "InBrainSurveys.h"
#import <InBrainSurveys_SDK_Swift/InBrainSurveys_SDK_Swift-Swift.h>

@implementation InBrainSurveys

// ***********************************
// ***** UIVIEWCONTROLER methods *****
// ***********************************
- (instancetype)init
{
    self = [super init];
    self.inbrain = [InBrain shared];
    [self.inbrain setInBrainDelegate:self];
    return self;
}

// *********************************
// ***** RN BRIDGE methods  ********
// *********************************

RCT_EXPORT_MODULE()

+ (BOOL)requiresMainQueueSetup
{
  return NO;  // only do this if your module initialization relies on calling UIKit!
}

// ***********************
// ***** SET INBRAIN *****
// ***********************
RCT_EXPORT_METHOD(setInBrain:(NSString *)apiClientId apiSecret:(nonnull NSString *)apiSecret isS2S:(BOOL)isS2S userId:(nonnull NSString *)userId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
        
        [self notNull:@"apiClientId" toCheck:apiClientId];
        [self notNull:@"apiSecret" toCheck:apiSecret];
        [self notNull:@"userId" toCheck:userId];

        [self.inbrain setInBrainWithApiClientID:apiClientId apiSecret:apiSecret isS2S:isS2S userID:userId];

        // Resolve
        resolve(nil);
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_INBRAIN", error.description, nil);
    }
}

// **********************************
// ***** SET INBRAIN VALUES FOR *****
// **********************************
RCT_EXPORT_METHOD(setInBrainValuesFor:(NSString *)sessionId data:(NSDictionary *)data resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {

        // Convert map to array of single entry maps
        // e.g {age: 25, gender: male} will become [{age: 25}, {gender: male}]
        NSArray* keys=[data allKeys];
        NSMutableArray *mapped = [NSMutableArray arrayWithCapacity:[keys count]];

        [keys enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            [mapped addObject:@{ obj : [data objectForKey:obj]}];
        }];

        [self.inbrain setInBrainValuesForSessionID:sessionId dataOptions:mapped];

        // Resolve
        resolve(nil);
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_INBRAIN_VALUES", error.description, nil);
    }
}

// ************************
// ***** SHOW SURVEYS *****
// ************************
RCT_EXPORT_METHOD(showSurveys:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    
    // This requires the main thread
    dispatch_async(dispatch_get_main_queue(), ^{
    
        @try {
            [self.inbrain showSurveysFrom:NULL];
            resolve(nil);
        }
        @catch (NSException *error) {
            reject(@"ERR_SHOW_SURVEYS", error.description, nil);
        }
    });
    
}

// ************************
// ***** GET REWARDS ******
// ************************
RCT_EXPORT_METHOD(getRewards:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
        [[InBrain shared] getRewardsWithSuccess:^(NSArray<InBrainReward *> * rewards){
            NSMutableArray *rewardList = [NSMutableArray array];
            for(int i = 0; i < rewards.count; i++) {

                // ENHANCE
                // The mapping seems to be necessary. Resolving the promise directly with 'rewards' array doesn't work
                // The result on the RN side is an array with null elements...
                 NSObject* o = @{@"transactionId": [NSNumber numberWithLong:rewards[i].transactionId], @"currency": rewards[i].currency, @"amount": [NSNumber numberWithDouble:rewards[i].amount], @"transactionType": [NSNumber numberWithFloat:rewards[i].transactionType]};

               [rewardList addObject:o];
            }


            resolve(rewardList);
        } failed:^(NSError * error){
            reject(@"ERR_GET_REWARDS", error.localizedDescription, error);
        }];
    }
    @catch (NSException *error) {
        reject(@"ERR_GET_REWARDS", error.description, nil);
    }

}

// ***********************************
// ***** CHECK SURVEYS AVAILABLE *****
// ***********************************
RCT_EXPORT_METHOD(checkSurveysAvailable:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    
    @try{

        [[InBrain shared] checkForAvailableSurveysWithCompletion:^(BOOL available, NSError * error) {
            if(error) reject(@"ERR_CHECK_SURVEYS_AVAILABLE", error.localizedDescription, nil);
            else resolve(@(available));
        }];

    }
    @catch (NSException *error) {
        reject(@"ERR_CHECK_SURVEYS_AVAILABLE", error.description, nil);
    }
}

// *******************************
// ***** GET NATIVE SURVEYS ******
// *******************************
RCT_EXPORT_METHOD(getNativeSurveys:(NSString * _Nullable)placementId resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
        [[InBrain shared] getNativeSurveysWithPlacementID:placementId success:^(NSArray<InBrainNativeSurvey *> * surveys){
            NSMutableArray *surveyList = [NSMutableArray array];
            for(int i = 0; i < surveys.count; i++) {

                // ENHANCE
                // The mapping seems to be necessary. Resolving the promise directly with 'surveys' array doesn't work
                // The result on the RN side is an array with null elements...
                 NSObject* o = @{@"id": surveys[i].id, @"rank": [NSNumber numberWithLong:surveys[i].rank], @"time": [NSNumber numberWithLong:surveys[i].time], @"value": [NSNumber numberWithDouble:surveys[i].value], @"currencySale": [NSNumber numberWithBool:surveys[i].currencySale], @"multiplier": [NSNumber numberWithDouble:surveys[i].multiplier]};

                   [surveyList addObject:o];
            }

            resolve(surveyList);
        } failed:^(NSError * failed){
            reject(@"ERR_GET_NATIVE_SURVEYS", failed.localizedDescription, failed);
        }];

    }
    @catch (NSException *error) {
        reject(@"ERR_GET_NATIVE_SURVEYS", error.description, nil);
    }
}

// *******************************
// ***** SHOW NATIVE SURVEY ******
// *******************************
RCT_EXPORT_METHOD(showNativeSurvey:(NSString*)id  placementId:(NSString*)placementId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try {
        // This requires the main thread
        dispatch_async(dispatch_get_main_queue(), ^{
            [[InBrain shared] showNativeSurveyWithId:id placementId:placementId from:NULL];
            resolve(@true);
        });
    }
    @catch (NSException *error) {
        reject(@"ERR_SHOW_NATIVE_SURVEY", error.description, nil);
    }
}

// ***************************
// ***** CONFIRM REWARDS *****
// ***************************
RCT_EXPORT_METHOD(confirmRewards:(NSArray *)rewards resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    
    @try{
        // Mapping to list of ids and forwarding to SDK
        NSArray* ids = [rewards valueForKey:@"transactionId"];
        [[InBrain shared] confirmRewardsWithTxIdArray:ids];

        // Resolve the promise
        resolve(@true);
    }
    @catch (NSException *error) {
        reject(@"ERR_CONFIRM_REWARDS", error.description, nil);
    }
}

// **************************
// ***** SET VIEW TITLE *****
// **************************
RCT_EXPORT_METHOD(setTitle:(NSString *)title resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        // Forwarding to SDK
        [[InBrain shared] setNavigationBarTitle:title];

        // Resolve the promise
        resolve(@true);
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_TITLE", error.description, nil);
    }
}

// *************************************
// ***** SET NAVIGATION BAR CONFIG *****
// *************************************
RCT_EXPORT_METHOD(setNavigationBarConfig:(NSDictionary *)data resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        
        // Extract parameters
        NSString* backgroundColorString = [data objectForKey:@"backgroundColor"];
        UIColor* backgroundColor = [self colorWithHexString:backgroundColorString];

        NSString* buttonsColorString = [data objectForKey:@"buttonsColor"];
        UIColor* buttonsColor = [self colorWithHexString:buttonsColorString];
        
        NSString* titleColorString = [data objectForKey:@"titleColor"];
        UIColor* titleColor = [self colorWithHexString:titleColorString];
        
        BOOL hasShadow = [[data objectForKey:@"hasShadow"] boolValue];
        
        // Instantiate config object
        InBrainNavBarConfig* config = [[InBrainNavBarConfig alloc] initWithBackgroundColor:backgroundColor buttonsColor:buttonsColor titleColor:titleColor isTranslucent:false hasShadow:hasShadow];

        // Forwarding to SDK
        [[InBrain shared] setNavigationBarConfig:config];

        // Resolve the promise
        resolve(@true);
    
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_NAVIGATION_BAR_CONFIG", error.description, nil);
    }
}

// *********************************
// ***** SET STATUS BAR CONFIG *****
// *********************************
RCT_EXPORT_METHOD(setStatusBarConfig:(NSDictionary *)data resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{

        // Extract parameters
        BOOL lighStatusBar = [[data objectForKey:@"lightStatusBar"] boolValue];
        
        UIStatusBarStyle style = 1;
        if(!lighStatusBar) {
            if(@available(iOS 13, *))
                style = 3; // UIStatusBarStyleDarkContent
            else
                style = 0; // UIStatusBarStyleDefault
        }
        
       // Instantiate config object
       InBrainStatusBarConfig* config = [[InBrainStatusBarConfig alloc] initWithStatusBarStyle:style hideStatusBar:false];
        
        // Forwarding to SDK
        [[InBrain shared] setStatusBarConfig:config];

        // Resolve the promise
        resolve(@true);
    
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_STATUS_BAR_CONFIG", error.description, nil);
    }
}

// ***********************
// ***** SET LANGUAGE ****
// ***********************
RCT_EXPORT_METHOD(setLanguage:(NSString *)language resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{

        // Forwarding to SDK
        NSError *error = nil;
        [[InBrain shared] setLanguage:language error:&error];
        if (error != nil) {
            reject(@"ERR_SET_LANGUAGE", error.description, nil);
        }

        // Resolve the promise
        resolve(@true);
    
    }
    @catch (NSException *error) {
        reject(@"ERR_SET_LANGUAGE", error.description, nil);
    }
}


// ********************
// ***** LISTENERS ****
// ********************

- (NSArray<NSString *> *)supportedEvents
{
  return @[@"OnClose", @"OnCloseFromPage"];
}

- (void)surveysClosedByWebView:(BOOL)byWebView completedSurvey:(BOOL)completedSurvey {
    [self sendEventWithName:(byWebView ? @"OnCloseFromPage" : @"OnClose") body:@{}];
}

- (void)didReceiveInBrainRewardsWithRewardsArray:(NSArray<InBrainReward *> * _Nonnull)rewardsArray {
    // Never used, we use getRewardsWithSuccess which has callbacks. This method is only used when getRewards is called.
}

// ***************************
// ***** UTILITY METHODS *****
// ***************************

/**
 * Convert from a hexadecimal color string to UIColor
 */
- (UIColor *)colorWithHexString:(NSString *)stringToConvert
{
    NSString *noHashString = [stringToConvert stringByReplacingOccurrencesOfString:@"#" withString:@""]; // remove the #
    NSScanner *scanner = [NSScanner scannerWithString:noHashString];
    [scanner setCharactersToBeSkipped:[NSCharacterSet symbolCharacterSet]]; // remove + and $

    unsigned hex;
    if (![scanner scanHexInt:&hex]) return nil;
    int r = (hex >> 16) & 0xFF;
    int g = (hex >> 8) & 0xFF;
    int b = (hex) & 0xFF;

    return [UIColor colorWithRed:r / 255.0f green:g / 255.0f blue:b / 255.0f alpha:1.0f];
}

- (void) notNull:(NSString* )name toCheck:(id)toCheck {
    if( !toCheck ){
        [NSException raise:@"Invalid parameter value" format:@"%@ must not be null", name];
    }
}

@end

